src/ImageSharp/Processing/Processors/Normalization/AdaptiveHistogramEqualizationProcessor{TPixel}.cs

```diff
@@ -146,23 +142,22 @@ protected override void OnFrameApply(ImageFrame<TPixel> source)
         {
             for (int dy = yStart; dy < yEnd; dy++)
             {
-                int dyOffSet = dy * sourceWidth;
```

Comment:
It seems to me like `int dyOffSet = dy * sourceWidth;` (_and the same below_) were the actual source of the issue and we could have mitigated the issue with rows whilekeeping `Unsafe.Add`.

I'd like to see a before/after benchmark if possible.

Reply:
> while keeping Unsafe.Add

Do we really want it? What I meant in https://github.com/SixLabors/ImageSharp/issues/1429#issuecomment-727931317 is a policy against `Unsafe.*` regardless of whether it can be justified in a particular use-case.

These processors are the spice in the library, it's not the core functionality where we really need the 1-2 extra percents got by micro-optimization. I'd rather go low-risk here. It's much better to get `IndexOutOfRangException` than an access violation which is a security-critical bug.

Reply:
I'd like to see benchmarks before I make a decision.

Reply:
I do agree we should be a lot more careful with `Unsafe` though. I'm guilty of reaching for it far too often.

Reply:
Think I've spotted a way to keep perf for all in a safe manner.

Reply:
Try a microbenchmark iterating through a span. You shall see no difference  or almost no difference if the loop is defined as `for (int i = 0; i < span.Length; i++)`.

Since real life code does much more than that, benefits shall be negligible or nonexistent. I barely see such code in BCL, I believe the only visible benefit is on .NET Framework, or maybe older .NET Core. 

If there is no proof for visible benefits, I would strongly prefer to see a guideline point against `Unsafe.*`.

Reply:
That's why I suggested the slice outside the loop in this instance so we can use `.Length`. There's little change in Core 3.1 with the current approach but Core 2.1 and NET 472 are suffering according to the benchmarks. 

>If there is no proof for visible benefits, I would strongly prefer to see a guideline point against Unsafe.*.

Yep. Totally with you. I know of several places in the library where it would be smart to revert to safer code.